src/python/grpcio_observability/grpc_observability/python_census_context.cc

```diff
@@ -0,0 +1,268 @@
+      header[kTraceOptionsOfs] != kTraceOptionsField) {
+    return SpanContext();  // Invalid.
+  }
+
+  uint8_t options = header[kTraceOptionsOfs + 1] & 1;
+  constexpr uint8_t kIsSampled = 1;
+
+  uint8_t trace_id_rep_[kTraceIdSize];
+  memcpy(trace_id_rep_, reinterpret_cast<const uint8_t*>(&header[kTraceIdOfs + 1]), kTraceIdSize);
+  
+  uint8_t span_id_rep_[kSpanIdSize];
+  memcpy(span_id_rep_, reinterpret_cast<const uint8_t*>(&header[kSpanIdOfs + 1]), kSpanIdSize);
+
+  uint8_t trace_option_rep_[kTraceOptionsLen];
+  memcpy(trace_option_rep_, &options, kTraceOptionsLen);
+
+  SpanContext context(absl::BytesToHexString(absl::string_view(reinterpret_cast<const char *>(trace_id_rep_), kTraceIdSize)),
+                      absl::BytesToHexString(absl::string_view(reinterpret_cast<const char *>(span_id_rep_), kSpanIdSize)),
+                      trace_option_rep_[0] & kIsSampled);
+  return context;
+}
+
+
+size_t TraceContextSerialize(PythonCensusContext& context,
+                             char* tracing_buf, size_t tracing_buf_size) {
+  if (tracing_buf_size < kGrpcTraceBinHeaderLen) {
+    return 0;
+  }
+  ToGrpcTraceBinHeader(context, reinterpret_cast<uint8_t*>(tracing_buf));
+  return kGrpcTraceBinHeaderLen;
+}
+
+
+size_t StatsContextSerialize(size_t /*max_tags_len*/, grpc_slice* /*tags*/) {
+  // TODO(unknown): Add implementation. Waiting on stats tagging to be added.
+  return 0;
+}
+
+
+size_t ServerStatsDeserialize(const char* buf, size_t buf_size,
+                              uint64_t* server_elapsed_time) {
+  return grpc::internal::RpcServerStatsEncoding::Decode(
+      absl::string_view(buf, buf_size), server_elapsed_time);
+}
+
+
+size_t ServerStatsSerialize(uint64_t server_elapsed_time, char* buf,
+                            size_t buf_size) {
+  return grpc::internal::RpcServerStatsEncoding::Encode(server_elapsed_time, buf,
+                                                  buf_size);
+}
+
+
+uint64_t GetIncomingDataSize(const grpc_call_final_info* final_info) {
+  return final_info->stats.transport_stream_stats.incoming.data_bytes;
+}
+
+
+uint64_t GetOutgoingDataSize(const grpc_call_final_info* final_info) {
+  return final_info->stats.transport_stream_stats.outgoing.data_bytes;
+}
+
+
+namespace {
+// span_id is a 16-character hexadecimal encoded string.
+std::string generateSpanId() {
+  uint64_t span_id = absl::Uniform<uint64_t>(absl::BitGen());
+  std::stringstream hex_string;
+  hex_string << std::setfill('0') << std::setw(16) << std::hex << span_id;
+  return std::string(hex_string.str());
+}
+
+// trace_id is a 32-character hexadecimal encoded string
+std::string generateTraceId() {
+  absl::uint128 trace_id = absl::Uniform<absl::uint128>(absl::BitGen());
+  std::stringstream hex_string;
+  hex_string << std::setfill('0') << std::setw(32) << std::hex << trace_id;
+  return std::string(hex_string.str());
+}
+
+} // namespace
+
+//
+// Span
+//
+
+Span Span::StartSpan(absl::string_view name, Span* parent) {
```

Comment:
You'll have to change around the `Span` APIs to make this work, but this should be `const Span*` since we're not mutating `parent` here.

To make this work, you'll need to add a `ConstContext()` method on `Span` or you could just get rid of the method-based accessor, make `Span::context_` public, and access it directly here.
 
More generally, if a method/function does not mutate one of its pointer/reference arguments (non-const value parameters are fine), then that parameter should be marked `const` unless we have a strong reason not to do so.